00-framework-01-maturity.md

```diff
@@ -0,0 +1,118 @@
+# libp2p specification framework – the five-stage maturity process
+
+> Author: @raulk
+
+> Revision: r0, 2019-05-21
+
+## Prelude
+
+This document establishes a framework to foster rapid and incremental libp2p
+specification development, aiming to reduce the barrier to entry for new
+ideas, increasing the throughput of breakthrough creative proposals, promoting
+their evolution and adoption within the community, while maximising consensus
+around a common lifecycle and criteria for progression across stages.
+
+We propose a five-stage process to classify the maturity of a specification in
+terms of completeness, demonstrability of implementation, community
+acceptance, and level of technical detail. It is somewhat inspired in the W3C
+Process [0].
+
+  1. `Working Draft`
+  2. `Candidate Recommendation`
+  3. `Recommendation`
+  4. `Deprecated`
+  5. `Abandoned Draft`
+
+## Stage 1: Working draft
+
+The specification of the system, process, protocol or item is under active
+development. This stage is lightweight and mostly self-directed by the author.
+We aim to reduce the barrier to entry, and it's designed to allow for
+iterative experimentation, discovery and pivoting.
+
+We don't enforce a hard template in an attempt to enhance author
+expressability and creativity.
+
+We enter this stage by posting an `Initial Working Draft` that covers:
+  * context: what is the current situation or a brief overview of the
+    environment the specification targets.
+  * motivation: why this specification is relevant, and how it advances the
+    status quo.
+  * scope and rationale: what areas of the technical system the specification
+    impacts.
+  * goals: what we expect to achieve (positively and negatively) as a result
+    of implementing the specification.
+  * expected feature set: a summary/enumeration of features the spec provides.
+  * tentative technical directions: how are we planning to materialise the
+    specification in terms of system design.
+
```

Comment:
Should we cover a checklist of expected points needed in order to get in stage 2? 

Examples: finish proof of work, implementation, ...

I think it would be valuable to have a clear notion of progress

Reply:
The way I see it, the author will need to enhance the spec so that it covers the requirements of a Candidate Recommendation:

```
The specification document itself is technically complete. It defines wire
level formats for interoperability, error codes, algorithms, data structures,
heuristics, behaviours, etc., in a way that it is sufficient to enable
contributors to develop an interoperable implementation.
```

And also there would need to be ONE conforming implementation.

At that stage, the author can submit a PR to promote the spec from Working Draft to Candidate Recommendation.

Reply:
My idea would be to allow the author to do an iterative work with several PRs, and the complete checklist would increase over time. 

With this, people checking the specs docs will have a better understanding of the current state of the spec without digesting all the content.

Reply:
I like the ability to track progress in that manner. I don't want to define a global checklist, though, because we'll very quickly find an exceptional case where the checklist is unfit. Perhaps a tradeoff is to have the author introduce a spec-relevant checklist during the working draft phase?

Reply:
Covered here: f65572f77dbee36638b1bfd63906b24cb8d61227. Let me know what you think.

Reply:
Yes, I agree with your proposal!